### ISSUE TITLE
maintain() has substantial overhead

### ISSUE THREAD

**Issue by m4b** [slide-rs/specs#555]
## Description

Calling maintain can have stunning amount of overhead, dwarfing dispatch system overhead.

## Meta

Rust version: nightly
Specs version / commit: 0.14.1 
Operating system: linux

## Reproduction

Create large amount of entities, with large amount of lazy updates.

## Expected behavior

I spent a large amount of time optimizing my systems to have minimal overhead, carefully watching performance, not allocating in loops etc., only to discover after profiling my systems in totality that a large amount of time (over half, sometimes 80-90% of loop time) was being spent elsewhere.

After investigating, I discovered (much to my disappointment) it was in the framework itself, over which I have little or no control.  Here is a sample profile:

```
nentities: 4371
  TOTAL      - 84.855505ms
  maintain   - 67.570179ms
  dispatch   - 17.121075ms
  ... system breakdown
  physics    - 6.80588ms
  foo        - 4.401008ms
  render     - 2.82132ms
  ...
```

To summarize, I have  4371 entities, a single loop takes 84ms, where a _whopping_ 67ms/84ms (80% !) is spent just in the `world.maintain()` call; dispatch is the sum of all my systems time, and the breakdown after is given as an example, the physics call being the largest.

And yes, this was compiled in release mode.

Is there a suggested pattern to prevent this kind of overhead? I haven't seen any documentation warning about these kind of shocking framework overheads, which is completely unworkable for my use case (which was tailored for 60fps). Am I just out of luck, or is there some optimization work that hasn't been done on that function call?

**Comment by m4b** [comment]
EDIT: the nentities I've given is incorrect; it is much much smaller, will try to get the exact number.

**Comment by Xaeroxe** [comment]
lazy updates will do that, they had to be run somewhere. If you don't want this to be a problem then don't do lazy updates. I'm sorry this answer is likely unsatisfactory, but I'd argue this isn't framework overhead, because you gave the framework a lot of work to do.

**Comment by m4b** [comment]
ok... yea it is pretty unsatisfying given that its basically undocumented that I can have an 80% overhead. I really feel like there is likely a highly unoptimized path somewhere in the updates causing that kind of latency, but I don't know the codebase that well, so perhaps not? But it seems pretty suspect.

Also perhaps I'm using the wrong pattern here, but I'm using lazy updates to iterate through each entity, and if condition X occurs, I lookup entity Z, and mutate that entity. Because the entities are already borrowed during the join, I use lazy updates to accomplish this; should I use another pattern here?

I guess I'll start combing through my various uses of lazy update and post them here for why I ended up using them, perhaps I can avoid it?

**Comment by m4b** [comment]
That being said, I find it hard to believe that I've lazily added **4x more work** in my systems (and again I'd suggest the function for resolving lazy work could probably be better optimized?). Is there a way for me to measure this easily that you know of?

**Comment by Xaeroxe** [comment]
It'd be helpful to have a minimal reproduction example. As is this isn't super actionable.

`maintain()` in most projects doesn't have this kind of execution time. Therefore I find it hard to believe this is framework overhead.

The way to avoid using lazy updates is to collect all the changes you need to make into a `Vec` during your iteration, and then make them in the same system. This allows your work to become parallel again. LazyUpdates, by nature of their design, have to execute one after another, costing you parallelism.

Here's an example


```rust
fn run(&mut self, (foo, bar): Self::SystemData) {
    let changes = Vec::new();
    for (foo, bar) in (&mut foo, &mut bar).join() {
        if foo.isFoo {
            changes.push(foo.id);
        }
    }
    for id in changes {
        // Do something
    }
}
```

If you still insist maintain has too much overhead you're welcome to try and optimize it. Here is where it lives: https://github.com/slide-rs/specs/blob/master/src/world/mod.rs#L596

And this is the code for the lazy updates specific piece. https://github.com/slide-rs/specs/blob/master/src/world/lazy.rs#L310

**Comment by Xaeroxe** [comment]
Oh and another optimization you can make to my code example is to store the vec allocation in the system structure so you can re-use it inbetween frames. That way you minimize the amount of time spent messing with the OS memory allocator.

```rust
struct MyCoolSystem {
    scratch_vec: Vec<Something>,
}

...

fn run(&mut self, (foo, bar): Self::SystemData) {
    self.scratch_vec.clear();
    for (foo, bar) in (&mut foo, &mut bar).join() {
        if foo.isFoo {
            changes.push(foo.id);
        }
    }
    for id in &self.scratch_vec {
        // Do something
    }
}

**Comment by Xaeroxe** [comment]
I'm going to go ahead and close this, if you feel I am in error feel free to re-open it.
